src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs

```diff
@@ -0,0 +1,598 @@
+
+            if (!_leaveOpen)
+            {
+                innerStream.Dispose();
+            }
+        }
+
+        public override ValueTask DisposeAsync()
+        {
+            if (_innerStream is null)
+            {
+                return default; // dispose called multiple times, ignore
+            }
+
+            // First, get any pending data destined for the inner stream.
+
+            ArraySegment<byte> pendingData = FinalFlushWriteBuffers();
+
+            if (pendingData.Count == 0)
+            {
+                // Fast path: just dispose of the object graph.
+                // No need to write anything to the stream first.
+
+                Stream innerStream = _innerStream;
+                _innerStream = null;
+
+                return (_leaveOpen)
+                    ? default /* no work to do */
+                    : innerStream.DisposeAsync();
+            }
+
+            // Slower path; need to perform an async write followed by an async dispose.
+
+            return DisposeAsyncCore(pendingData);
+            async ValueTask DisposeAsyncCore(ArraySegment<byte> pendingData)
+            {
+                Debug.Assert(pendingData.Count != 0);
+
+                Stream innerStream = _innerStream;
+
+                await innerStream.WriteAsync(pendingData.AsMemory()).ConfigureAwait(false);
+                _innerStream = null;
+
+                if (!_leaveOpen)
+                {
+                    await innerStream.DisposeAsync().ConfigureAwait(false);
+                }
+            }
+        }
+
+        public override int EndRead(IAsyncResult asyncResult)
+            => TaskToApm.End<int>(asyncResult);
+
+        public override void EndWrite(IAsyncResult asyncResult)
+            => TaskToApm.End(asyncResult);
+
+#pragma warning disable CS3016 // Arrays as attribute arguments is not CLS-compliant
+#pragma warning disable CS8774 // Member must have a non-null value when exiting.
+
+        // Sets up the data structures that are necessary before any read operation takes place,
+        // throwing if the object is in a state where reads are not possible.
+        [MemberNotNull(nameof(_innerStream), nameof(_innerDecoder), nameof(_thisEncoder))]
+        private void EnsurePreReadConditions()
+        {
+            ThrowIfDisposed();
+            if (_innerDecoder is null)
+            {
+                InitializeReadDataStructures();
+            }
+
+            void InitializeReadDataStructures()
+            {
+                if (!CanRead)
+                {
+                    throw Error.GetReadNotSupported();
+                }
+
+                _innerDecoder = _innerEncoding.GetDecoder();
+                _thisEncoder = _thisEncoding.GetEncoder();
+                _readCharBufferMaxSize = _innerEncoding.GetMaxCharCount(DefaultReadByteBufferSize);
+
+                // Can't use ArrayPool for the below array since it's an instance field of this object.
```

Comment:
We don't follow this requirement in lots of other places.

Reply:
As instance fields on a public non-thread safe class? That's surprising - I thought we were fairly consistent about avoiding that.